### PR TITLE
dialects: (llvm) add void and function types

### PR DESCRIPTION
**Pull request by superlopuh**

### CODE REVIEW

File: xdsl/dialects/llvm.py [hunk 1]
```diff
@@ -185,6 +184,63 @@ def from_size_and_type(size: int | IntAttr, type: Attribute):
...
+@irdl_attr_definition
+class LLVMVoidType(ParametrizedAttribute, TypeAttribute):
+    name = "llvm.void"
+
+    def print_parameters(self, printer: Printer) -> None:
+        ...
+
+    @staticmethod
+    def parse_parameters(parser: Parser) -> list[Attribute]:
+        return []
```
Comment (math-fehr) [inline xdsl/dialects/llvm.py:196]: What happens if you don't specify it? It should still work, right?

**Review by georgebisbas — APPROVED** [review]
Unless other tests are not needed for pytest I am fine

**Review by math-fehr — APPROVED** [review]
Nice! Can you add an interoperation test with MLIR as well?

File: xdsl/dialects/llvm.py (hunk 1 above)
Comment (superlopuh) [reply xdsl/dialects/llvm.py:196]: Ah no idea, I thought it was required, let me check

**Comment by AntonLydike** [comment]
Can we get interop filechecks for this? I think it's important that we keep the llvm dialect interop going...

**Review by webmiche — COMMENTED** [review]
MLIR interoperation tests? ;) edit: nvm Anton was first again :(

**Comment by superlopuh** [comment]
@webmiche wrote my first interop tests, do you think they're good to merge?

File: tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// RUN: mlir-opt %s --mlir-print-op-generic | xdsl-opt --print-op-generic | filecheck %s
+
+"builtin.module"() ({
+  // CHECK:   "builtin.module"() ({
+
+  // Type tests
+
+  // literal
+  %s0, %s1 = "test.op"() : () -> (!llvm.struct<()>, !llvm.struct<(i32)>)
```
Comment (webmiche) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir:9]: Wait you can pass test ops through mlir? Without allowing unregistered dialects? I am confused :thinking:

**Review by webmiche — APPROVED** [review]
Lgtm

File: tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir (hunk 2 above)
Comment (superlopuh) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir:9]: The test seems to pass, so I guess it works 🤷

Comment (webmiche) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir:9]: Hmm weird. We should investigate that.

Maybe add an issue? I think we can merge this meanwhile.

File: xdsl/dialects/llvm.py [hunk 3]
```diff
@@ -185,6 +184,56 @@ def from_size_and_type(size: int | IntAttr, type: Attribute):
...
+
+    inputs: ParameterDef[ArrayAttr[Attribute]]
+    output: ParameterDef[Attribute]
+
+    def __init__(
+        self, inputs: Sequence[Attribute] | ArrayAttr[Attribute], output: Attribute
+    ) -> None:
+        if not isinstance(inputs, ArrayAttr):
+            inputs = ArrayAttr(inputs)
+        super().__init__([inputs, output])
```
Comment (AntonLydike) [inline xdsl/dialects/llvm.py:208]: Maybe we can make output type optional and automatically assign void here?

**Review by AntonLydike — APPROVED** [review]
Very nice!

**Comment by AntonLydike** [comment]
I guess this doesn't support variadics?

```
!llvm.func<void (i32, ...)>   // a variadic function with at least one argument.
```

Maybe add a quick comment to the op that that's not supported, otherwise it may get confusing?

**Comment by superlopuh** [comment]
yeah, not yet, I'll add the comment
- [ ] add comment
- [ ] implicit null result

File: tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir (hunk 2 above)
Comment (superlopuh) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/llvm/llvm_types.mlir:9]: #1018

**Pull request merged**